Set log level from env var FLB_LOG_LEVEL

PettitWesley opened the pull request:
#1769 

Previously, log level could be set with the following precedence:
1. Via `Log_Level` in the [SERVICE] section of a config file
2. Command Line args `-q` and `-v` 

Now the options and precedence are:
1. Value of the Env Var `FLB_LOG_LEVEL`
2. Via `Log_Level` in the [SERVICE] section of a config file
3. Command Line args `-q` and `-v`

PettitWesley commented:
Work in progress until I figure out how to best prevent the warning [here](https://github.com/fluent/fluent-bit/blob/master/src/flb_env.c#L146) from being printed:

```
[ warn] [env] variable ${FLB_LOG_LEVEL} is used but not set
```

This occurs when the env var is not set.

PettitWesley commented:
## Manual Testing

Base Config:

```
[INPUT]
    Name cpu
    Tag cpu

[OUTPUT]
    Name stdout
    Match *
```

### Existing Behavior

##### Default settings: log level defaults to info ✅

```
$ ./bin/fluent-bit -c ~/log-level/fluent-bit.conf
Fluent Bit v1.4.0
Copyright (C) Treasure Data

[2019/11/29 07:19:36] [ info] [storage] initializing...
[2019/11/29 07:19:36] [ info] [storage] in-memory
[2019/11/29 07:19:36] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2019/11/29 07:19:36] [ info] [engine] started (pid=27801)
[2019/11/29 07:19:36] [ info] [sp] stream processor started
```

##### Set log level in config file

Add `Log_Level debug` in Service section.

```
```

##### Set log level in config file via an env var

Add `Log_Level ${LEVEL}` in Service section.

##### Set log level with `-v` flag

##### Disable logging with `-q` flag

### New Behavior

##### Set log level with `FLB_LOG_LEVEL`

##### Set log level with `FLB_LOG_LEVEL` and override config file

##### Set log level with `FLB_LOG_LEVEL` and override `-v`

##### Set log level with `FLB_LOG_LEVEL` and override `-q`

PettitWesley commented:
No longer WIP. This is ready for review.

edsiper commented:
thanks for your contribution!

from a usability perspective, why do we need to add a fixed environment variable for a specific flag from the main configuration section ?, I think if we introduce this, shortly we will have to add more environment variables for every supported variable by the configuration.

At the moment to use an environment variable in the config is enough using the current format through [Fluent Bit variables](https://docs.fluentbit.io/manual/configuration/variables)  without the need to add a new way to configure it:

```
$ cat fluent-bit.conf 
[SERVICE]
    Log_Level  ${MYLOGLEVEL}

[INPUT]
    Name cpu

[OUTPUT]
    Name  stdout
    Match *
[@monox:~/coding/fluent-bit/build GIT(context) 0]
$ MYLOGLEVEL=debug bin/fluent-bit -c fluent-bit.conf 
Fluent Bit v1.4.0
Copyright (C) Treasure Data

[2019/11/30 00:13:52] [debug] [storage] [cio stream] new stream registered: cpu.0
[2019/11/30 00:13:52] [ info] [storage] initializing...
[2019/11/30 00:13:52] [ info] [storage] in-memory
[2019/11/30 00:13:52] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2019/11/30 00:13:52] [ info] [engine] started (pid=28721)
[2019/11/30 00:13:52] [debug] [engine] coroutine stack size: 24576 bytes (24.0K)
[2019/11/30 00:13:52] [debug] [router] match rule cpu.0:stdout.0
[2019/11/30 00:13:52] [ info] [sp] stream processor started
```

PettitWesley commented:
>     Log_Level  ${MYLOGLEVEL}

I don't love that though; sorry. That means you have to always set a value for your env var, otherwise Fluent Bit will print a warning. Adding that setting kind of becomes boilerplate that you have to add in every config file. 

I think it's simpler just to have a standard env var.

nigels-com commented:
@edsiper The use case I can think of in support of this change:

The dev team have pushed a docker container to a repository (such as docker hub) along with some instructions about usage options.  The operations team are noticing some unexpected runtime behavior and wish to troubleshoot and narrow the (likely operational) problem.  The operations team might not know where the container came from, or the precise process of building a verbose-logging version of the same container, or how to prevent that being deployed more widely.  But if they can restart a container in-place with **docker run -e "FLB_LOG_LEVEL=debug"** they're quite likely to get the perspective they need with a minimum of fuss.

I don't think it's a slippery slope of needing to override _everything_ via environment variables, but logging would be one of a handful that are likely to be valuable for troubleshooting.

nigels-com commented:
Looks good to me.  I'd be willing to repeat and confirm the test plan locally, if it helps advance this.

nigels-com commented:
Just by chance I was poking around in a [systemd](https://www.freedesktop.org/software/systemd/man/systemd.html) service configuration that also supports environment variables along side the command-line.

It would be quite nice in this case to have `# Environment="FLB_LOG_LEVEL=debug"` as a convenient toggle, for troubleshooting purposes.  This would not require changing (and reliablly undoing) changes to the command-line, or configuration.

```
$ cat /etc/systemd/system/temper-python.service 
[Unit]
Description=temper-python service using fluent-bit and temper-python
After=network.target
StartLimitIntervalSec=0

[Service]
Type=simple
Restart=always
RestartSec=1
User=root
ExecStart=/usr/local/bin/fluent-bit -c /usr/local/etc/fluent-bit/temper-python.cfg -R /usr/local/etc/fluent-bit/parsers.conf

[Install]
WantedBy=multi-user.target
```

nigels-com commented:
Yep, I tested this branch with systemd and `Environment="FLB_LOG_LEVEL=debug"`.  Flawless.

```
Dec 12 11:51:08 pi1 systemd[1]: Started temper-python service using fluent-bit and temper-python.
Dec 12 11:51:08 pi1 fluent-bit[5529]: Fluent Bit v1.4.0
Dec 12 11:51:08 pi1 fluent-bit[5529]: Copyright (C) Treasure Data
Dec 12 11:51:08 pi1 fluent-bit[5529]: [2019/12/12 11:51:08] [debug] [storage] [cio stream] new stream registered: exec.0
Dec 12 11:51:08 pi1 fluent-bit[5529]: [2019/12/12 11:51:08] [debug] [storage] [cio stream] new stream registered: cpu.1
Dec 12 11:51:08 pi1 fluent-bit[5529]: [2019/12/12 11:51:08] [debug] [storage] [cio stream] new stream registered: mem.2
Dec 12 11:51:08 pi1 fluent-bit[5529]: [2019/12/12 11:51:08] [debug] [storage] [cio stream] new stream registered: thermal.3
Dec 12 11:51:08 pi1 fluent-bit[5529]: [2019/12/12 11:51:08] [ info] [storage] initializing...
Dec 12 11:51:08 pi1 fluent-bit[5529]: [2019/12/12 11:51:08] [ info] [storage] in-memory
Dec 12 11:51:08 pi1 fluent-bit[5529]: [2019/12/12 11:51:08] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
Dec 12 11:51:08 pi1 fluent-bit[5529]: [2019/12/12 11:51:08] [ info] [engine] started (pid=5529)
Dec 12 11:51:08 pi1 fluent-bit[5529]: [2019/12/12 11:51:08] [debug] [engine] coroutine stack size: 24576 bytes (24.0K)
Dec 12 11:51:08 pi1 fluent-bit[5529]: [2019/12/12 11:51:08] [debug] [in_exec] interval_sec=60 interval_nsec=0
Dec 12 11:51:09 pi1 fluent-bit[5529]: [2019/12/12 11:51:09] [debug] [router] match rule exec.0:stdout.0
Dec 12 11:51:09 pi1 fluent-bit[5529]: [2019/12/12 11:51:09] [debug] [router] match rule exec.0:http.1
Dec 12 11:51:09 pi1 fluent-bit[5529]: [2019/12/12 11:51:09] [debug] [router] match rule cpu.1:stdout.0
Dec 12 11:51:09 pi1 fluent-bit[5529]: [2019/12/12 11:51:09] [debug] [router] match rule cpu.1:http.1
Dec 12 11:51:09 pi1 fluent-bit[5529]: [2019/12/12 11:51:09] [debug] [router] match rule mem.2:stdout.0
Dec 12 11:51:09 pi1 fluent-bit[5529]: [2019/12/12 11:51:09] [debug] [router] match rule mem.2:http.1
Dec 12 11:51:09 pi1 fluent-bit[5529]: [2019/12/12 11:51:09] [debug] [router] match rule thermal.3:stdout.0
Dec 12 11:51:09 pi1 fluent-bit[5529]: [2019/12/12 11:51:09] [debug] [router] match rule thermal.3:http.1
Dec 12 11:51:09 pi1 fluent-bit[5529]: [2019/12/12 11:51:09] [ info] [sp] stream processor started
Dec 12 11:51:46 pi1 fluent-bit[5529]: [engine] caught signal (SIGTERM)
Dec 12 11:51:46 pi1 fluent-bit[5529]: [2019/12/12 11:51:46] [ info] [input] pausing exec.0
Dec 12 11:51:46 pi1 fluent-bit[5529]: [2019/12/12 11:51:46] [ info] [input] pausing cpu.1
Dec 12 11:51:46 pi1 fluent-bit[5529]: [2019/12/12 11:51:46] [ info] [input] pausing mem.2
Dec 12 11:51:46 pi1 fluent-bit[5529]: [2019/12/12 11:51:46] [ info] [input] pausing thermal.3
Dec 12 11:51:46 pi1 fluent-bit[5529]: [2019/12/12 11:51:46] [ warn] [engine] service will stop in 5 seconds
Dec 12 11:51:46 pi1 systemd[1]: Stopping temper-python service using fluent-bit and temper-python...
Dec 12 11:51:51 pi1 fluent-bit[5529]: [2019/12/12 11:51:51] [ info] [engine] service stopped
Dec 12 11:51:51 pi1 fluent-bit[5529]: [2019/12/12 11:51:51] [ info] [input] pausing exec.0
Dec 12 11:51:51 pi1 fluent-bit[5529]: [2019/12/12 11:51:51] [ info] [input] pausing cpu.1
Dec 12 11:51:51 pi1 fluent-bit[5529]: [2019/12/12 11:51:51] [ info] [input] pausing mem.2
Dec 12 11:51:51 pi1 fluent-bit[5529]: [2019/12/12 11:51:51] [ info] [input] pausing thermal.3
Dec 12 11:51:51 pi1 systemd[1]: Stopped temper-python service using fluent-bit and temper-python.
```

nigels-com commented:
Two minor issues/suggestions:

1. It would be nice to have a way to silence logging such as `FLB_LOG_LEVEL=off`

2. The result of `FLB_LOG_LEVEL=error` is a bit strange, warning seems to work fine.

```
$ FLB_LOG_LEVEL=error fluent-bit -c test.cfg 
Fluent Bit v1.4.0
Copyright (C) Treasure Data

[2019/12/12 12:05:55] [ info] Configuration
[2019/12/12 12:05:55] [ info]  flush time     : 0 seconds
[2019/12/12 12:05:55] [ info]  input plugins  : 
[2019/12/12 12:05:55] [ info] cpu 
[2019/12/12 12:05:55] [ info]  collectors     :
```

edsiper commented:
Understood, I am sold :) 

@nigels-com do you think the PR is good to go?

PettitWesley commented:
@nigels-com @edsiper 

> The result of FLB_LOG_LEVEL=error is a bit strange,

I've seen that too. For me it happened whenever I set log level to error, regardless of if it was with the env var or config file. 

I can try to track down where those statements are coming from and fix it in a separate PR (I don't believe this change introduced that).

edsiper commented:
@nigels-com do you think this is ready to go ?

nigels-com commented:
I'd be happy to do some further testing.  Caught up with pre-Christmas goings on lately.

nigels-com commented:
I've taken a fresh and detailed look at this changeset.

I do have one minor suggestion.  Both calls to `set_log_level_from_env` ought to be wrapped with a compile-time check for `#ifndef FLB_HAVE_STATIC_CONF` in my opinion.  The [documentation](https://docs.fluentbit.io/manual/installation/build_static_configuration) states:

> Static configuration mode aims to include a built-in configuration in the final binary of Fluent Bit, disabling the usage of external files or flags at runtime.

I also repeated the test plan from above, looks good to me.  I'd be happy to follow up with separate branches for the `FLB_LOG_LEVEL=Error` and `FLB_HAVE_STATIC_CONF` corner cases.

PettitWesley commented:
@nigels-com and @edsiper I've added the check for `FLB_HAVE_STATIC_CONF `. I think this is good to go!

I've also fixed the weird info messages when log level is error in a separate PR: https://github.com/fluent/fluent-bit/pull/1921

edsiper commented:
thanks for this.

In order to merge it please split the commits per component:

- env: ...
- log: ...
- bin: ...

edsiper commented:
thanks!
